packages/arb-token-bridge-ui/src/hooks/CCTP/useApproveAndDeposit.ts

```diff
@@ -0,0 +1,90 @@
+import { Provider } from '@ethersproject/providers'
+import { Signer, utils } from 'ethers'
+import { useCallback } from 'react'
+import { useToken } from 'wagmi'
+import { getTokenAllowanceForSpender } from '../../util/TokenUtils'
+import { useCCTP, UseCCTPParams, getContracts } from './useCCTP'
+
+export function useApproveAndDeposit({
+  sourceChainId,
+  walletAddress
+}: UseCCTPParams) {
+  const { approveForBurn, depositForBurn } = useCCTP({
+    sourceChainId,
+    walletAddress
+  })
+  const { usdcContractAddress, tokenMessengerContractAddress } =
+    getContracts(sourceChainId)
+  const { data: usdcToken } = useToken({ address: usdcContractAddress })
+  const approveAndDepositForBurn = useCallback(
+    async ({
+      amount,
+      provider,
+      signer,
+      destinationAddress,
+      onAllowanceTooLow = async () => {
+        return true
+      },
+      onApproveTxFailed,
+      onDepositTxFailed
+    }: {
+      amount: string
+      provider: Provider
+      signer: Signer
+      destinationAddress: `0x${string}` | undefined
+      onAllowanceTooLow: () => Promise<boolean>
```

Comment:
How come `onAllowanceTooLow` returns a `boolean`? Also, might be useful to return the current allowance:

```suggestion
      onAllowanceTooLow: (allowance: BigNumber) => Promise<void>
```

Reply:
i think this has to be renamed
`onAllowanceTooLow` sounds like a callback fn
maybe call it `isAllowanceTooLow`?

Reply:
`onAllowanceTooLow` is indeed a callback. When an user call `approveAndDepositForBurn`, we have this callback so they can for example display a modal and choose or not to proceed (based on the return boolean). But passing the allowance to the callback is a good idea
https://github.com/OffchainLabs/arbitrum-token-bridge/pull/1050/files#diff-286a4ee4334ef698f2f4080f95722061ff721facd19401fdd32279bf686c1e91R549-R554 (file is not expanded by default)

Reply:
Still don't see why we need to declare it as `Promise<boolean>` for the return value. The callback provider should be able do whatever they want, no?